libraries/AP_HAL_ChibiOS/RCOutput.cpp

```diff
@@ -2129,44 +2197,59 @@ void RCOutput::serial_bit_irq(void)
 void RCOutput::serial_byte_timeout(virtual_timer_t* vt, void *ctx)
 {
     chSysLockFromISR();
-    irq.timed_out = true;
-    chEvtSignalI((thread_t *)ctx, serial_event_mask);
+    uint16_t byteval = irq.bitmask | (((1U<<BYTE_BITS)-1) & ~((1U<<irq.nbits)-1));
+    // we can accept a byte with a timeout if the last bit was 1
+    // and the start bit is set correctly
+    if (irq.last_bit == 0) {
+        byteval = BAD_BYTE;
+    } else if ((byteval & 0x201) != 0x200) {
+        // wrong start/stop bits
+        byteval = BAD_BYTE;
+    }
+
+    serial_buffer.write((uint8_t*)&byteval, 2);
     chSysUnlockFromISR();
+    serial_sem.signal_ISR();
 }
 
 /*
   read a byte from a port, using serial parameters from serial_setup_output()
 */
-bool RCOutput::serial_read_byte(uint8_t &b)
+bool RCOutput::serial_read_byte(uint8_t &b, uint32_t timeout_us)
 {
-    irq.timed_out = false;
-    chVTSet(&irq.serial_timeout, chTimeMS2I(10), serial_byte_timeout, irq.waiter);
-    bool timed_out = ((chEvtWaitAny(serial_event_mask) & serial_event_mask) == 0) || irq.timed_out;
+    while (true) {
+        // consumer/producer pattern
+        if (serial_buffer.is_empty()) {
+            if (!serial_sem.wait(timeout_us)) {
+                return false;  // no data after 2ms
+            }
+        }
 
-    uint16_t byteval = irq.byteval;
+        chSysLock();
+        if (serial_buffer.is_empty()) {
+            chSysUnlock();
+            continue;
+        }
 
-    if (timed_out) {
-        // we can accept a byte with a timeout if the last bit was 1
-        // and the start bit is set correctly
-        if (irq.last_bit == 0) {
+        uint16_t byteval;
+        serial_buffer.read((uint8_t*)&byteval, 2);
```

Comment:
What says there are two bytes in there?  You're not checking the return value....

Reply:
We only ever write two bytes and we just checked that its not empty

Reply:
but I fixed it anyway